feat: introduce feature drag and pointer events handling

gabbopalma opened the pull request:
Closes #151 issue.

rmmlopes commented:
Hi, I'm working on a project that needs to update polygons on the map.  This PR would make that possible.  

Any idea when it will be merged in?

gabbopalma commented:
> Hi, I'm working on a project that needs to update polygons on the map.  This PR would make that possible.  
> 
> 
> 
> Any idea when it will be merged in?

Hi. I have currently stopped work because of the Christmas holidays. From January 7 I will resume work on this PR. I am close to the end of the work

rmmlopes commented:
> > Hi, I'm working on a project that needs to update polygons on the map.  This PR would make that possible.
> > Any idea when it will be merged in?
> 
> Hi. I have currently stopped work because of the Christmas holidays. From January 7 I will resume work on this PR. I am close to the end of the work

Awesome!  Thanks!

gabbopalma commented:
I decided not to introduce the Web implementation in the current PR.
The equality problems encountered in switching from GeometryCollection to FeatureCollection have been solved.

Implementation of a new way of handling gestures and, consequently, iOS and Web support will follow in a future PR.
At the moment, if @josxha approves it, I will introduce the “Android only works” warning in the code.

josxha commented:
Hi @gabbopalma, I'm sorry for your long wait. I've currently a long going on and probably won't find the time until February for an in-depth review. Though, I did a quick scroll-though and your pr looks promising! 
ktlint has a command to fix the formatting issues automatically but I can do it as you requested (can't remember the command at the moment).

gabbopalma commented:
I tested my code on a potential use case and I think a filter for interactive layers needs to be added.
Currently all features are detected, even those of `style.json` and this may be useless for most use cases for which `MapEventFeatureDrag` is used (like for moving and editing features).
We need to decide how and where to implement the interactive layers field @josxha

josxha commented:
Not sure why codecov lost its configuration. 👀 

> We need to decide how and where to implement the interactive layers field

Could you elaborate a bit please? I'm currently not up-to-date. The queryRenderedFeatures has a filter, hasn't it? It the question how to make use of it when dragging a feature?

> Also, I ran ktlintformat but I used version 0.4.17 and I think the version of ktlint is different from the pipeline used. Do we want to update it, or do I try to format it as the current version used?

It should use the following ruleset:
https://github.com/josxha/flutter-maplibre/blob/85a3d7bfee62a5c210e20927fe9cf8f2c102f5c7/android/build.gradle#L107
Maybe the dependencies were not loaded?

gabbopalma commented:
> > We need to decide how and where to implement the interactive layers field
> 
> Could you elaborate a bit please? I'm currently not up-to-date. The queryRenderedFeatures has a filter, hasn't it? It the question how to make use of it when dragging a feature?

Yes, queryRenderedFeatures has a filter, but it is not currently used. I was thinking if we want the user to pass to MapLibreMap, or somewhere else, the list of layers that will then be queried by the MapEventFeatureDrag.
Currently, the fastest solution, but with less customization, would be to pass all the layers that are present in the LayerManager, but this wouldn't make the features present in the layers added with StyleLayer queryable.


> Maybe the dependencies were not loaded?

I will try again. Maybe my Android Studio version has some conflicts with this. I was using it to run the gradle commands.

josxha commented:
I see. Yes, I agree it could be good to have fine control over what layers can get dragged and what can't. I've seen that the it is currently defined in `MapOptions.gestures`. How about moving this setting to `Layer.draggable`? Maybe even an enum that allows to specify the mode (on short press, on a long press, disabled). Would it work that way?

gabbopalma commented:
> I see. Yes, I agree it could be good to have fine control over what layers can get dragged and what can't. I've seen that the it is currently defined in `MapOptions.gestures`. How about moving this setting to `Layer.draggable`? Maybe even an enum that allows to specify the mode (on short press, on a long press, disabled). Would it work that way?

Yes. I think Layer.draggable (or even in StyleLayer) may be the best way to handle this. I will implement it.

gabbopalma commented:
I think I have finished my work here. I tried adding some tests for better coverage, but codecov still has a lower percentage than the target. Maybe we can collaborate on this (?)

I will wait for your review :)

josxha reviewed (commented):
Hi @gabbopalma, sorry for your long wait. You pull request contains a lot of good stuff. Can you check my comments please?

Could you give me some directions for testing, please? I wasn't able to test the drag functionality in the example app. Is there an example I can use for testing?

I rebased the pr to a new `next` branch for the v0.3.0 release. WIth the release of Flutter 3.29.0 I'd like to sneak in another feature release.

Thanks for your work!

ebw-gabrielpalmisano commented on android/src/main/kotlin/com/github/josxha/maplibre/MapLibreMapController.kt:
```diff
@@ -389,4 +404,54 @@ class MapLibreMapController(
+            flutterApi.onLongPressMove(LongPressEventType.END, point) {}
+
+            // Reset the move gesture listener to the default one.
+            mapLibreMap.setGesturesManager(defaultGesturesManager, true, true)
```
Yes. It is not the gesture manager that changes during a drag, but its listener changes and therefore the gesture manager also changes. 
So, I did not find a way to restore the default listener and decided to restore the gesture handler with its default value saved at the onMapReady start.

gabbopalma replied:
Yes. It is not the gesture manager that changes during a drag, but its listener changes and therefore the gesture manager also changes.
So, I did not find a way to restore the default listener and decided to restore the gesture handler with its default value saved at the onMapReady start.

gabbopalma commented on android/src/main/kotlin/com/github/josxha/maplibre/MapLibreMapController.kt:
```diff
@@ -389,4 +404,54 @@ class MapLibreMapController(
+
+            // Reset the move gesture listener to the default one.
+            mapLibreMap.setGesturesManager(defaultGesturesManager, true, true)
+            mapLibreMap.uiSettings.setAllGesturesEnabled(true)
```
Effectively this could have led to problems within the disabled gestures.
I inserted a system via a class, which handles the backup and restore of the UiSettings.

gabbopalma commented:
Hi @josxha. Thank you for your review :)
As requested, I fixed all the issues you had reported within the comments, and while I was at it I applied the following changes:
- Refactor of the StyleController and implementation of its draggableLayers. I preferred a method rather than a list allocated within the interface.
That way we have more control over the getter and setter of that list.
- Add examples in app regarding interactive Layers and StyleLayers.
- Added new “layerId” and “sourceId” fields within Layer and to all its subclasses.
This allows the override of getLayerId and getSourceId, in case the developer wants to apply a specific id.

josxha commented:
Thanks @gabbopalma👍

> I preferred a method rather than a list allocated within the interface. That way we have more control over the getter and setter of that list.
I agree, that it makes sense from setting the instance but in general not. Ihe returned instance is completely mutable and can get altered. A final field would have the same use. If the field needs to be updatable, a dart getter would work as well.

> Add examples in app regarding interactive Layers and StyleLayers.
Thanks! Is it showcasing the events or is it already somehow possible to drag markers?

> Added new “layerId” and “sourceId” fields within Layer and to all its subclasses.
This allows the override of getLayerId and getSourceId, in case the developer wants to apply a specific id.

I think this change is spot on. Enfording one (or both) of these fields at some point in the future might be worth. It shouldn't be a big deal for a user to add a unique and recognizable id. 🤔

Let me know when you want to have this pull request merged to the repo.

gabbopalma commented:
Hi @josxha, thank you for the PR review :)
Currently, the examples show the ability to drag layers and StyleLayers, in both categories. Actually, they don't have a legend, like with the existing example for Interactive Widgets example, if we want we can add it before pushing to main.

Regarding the use case of the unique id on layer or source, it is my particular need, but it could serve others for querying and managing layers with a certain ID pattern.

I would prefer the PR to be united as soon as possible, if possible ;)

josxha reviewed (approved):
Alright let's merge. Thanks for another contribution @gabbopalma 👍

rmmlopes commented:
Thank you @gabbopalma for this feature.  

Are you planning on enabling it on web and ios?  If not, whats the plan for doing that?

gabbopalma commented:
Hi @rmmlopes, it is currently not on the agenda, but I will think about this implementation soon. I need to do some testing on it first.
